### PR TITLE
move: tidy lock file header read interface

### PR DESCRIPTION
**Pull request by rvantonder**
## Description 

Just a cleanup of how we do lock file header reads. See separate commits for a breakdown of the change, and inline comments for more.

Follow up of https://github.com/MystenLabs/sui/pull/15751#pullrequestreview-1829957061. 

## Test Plan 

Should not be changing anything functionally, covered by existing tests.

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/crates/move-package/src/lock_file/schema.rs [hunk 1]
```diff
@@ -133,32 +133,32 @@ impl ToolchainVersion {
...
         };
-        read_header(&contents)
+        Self::from_str(&contents)
     }
-}
 
-/// Read lock file header after verifying that the version of the lock is not newer than the version
-/// supported by this library.
-#[allow(clippy::ptr_arg)] // Allowed to avoid interface changes.
-pub fn read_header(contents: &String) -> Result<Header> {
```
Comment (rvantonder) [inline external-crates/move/crates/move-package/src/lock_file/schema.rs:149]: main change: this free function goes away, becomes part of `Header` impl, and not `pub` scope any more.

File: external-crates/move/crates/move-package/src/resolution/dependency_graph.rs [hunk 2]
```diff
@@ -201,8 +202,10 @@ impl<Progress: Write> DependencyGraphBuilder<Progress> {
         // compute digests eagerly as even if we can't reuse existing lock file, they need to become
         // part of the newly computed dependency graph
         let new_manifest_digest = digest_str(manifest_string.into_bytes().as_slice());
+        let lock_path = root_path.join(SourcePackageLayout::Lock.path());
+        let mut lock_file = File::open(lock_path)?;
         let (old_manifest_digest_opt, old_deps_digest_opt, lock_string) = match lock_string_opt {
-            Some(lock_string) => match schema::read_header(&lock_string) {
+            Some(lock_string) => match schema::Header::read(&mut lock_file) {
```
Comment (rvantonder) [inline external-crates/move/crates/move-package/src/resolution/dependency_graph.rs:208]: This is the only place we did `schema:read_header` outside of the lock file schema, and now it uses `schema::Header::read`, which now becomes the canonical public interface for reading the lock file header.

File: external-crates/move/crates/move-package/src/lock_file/schema.rs [hunk 3]
```diff
@@ -106,7 +106,7 @@ impl Packages {
         let Schema { move_: packages } =
             toml::de::from_str::<Schema<Packages>>(&contents).context("Deserializing packages")?;
 
-        Ok((packages, read_header(&contents)?))
+        Ok((packages, Header::from_str(&contents)?))
```
Comment (rvantonder) [inline external-crates/move/crates/move-package/src/lock_file/schema.rs:109]: This is the only place we did schema:read_header outside of the lock file schema, and now it uses schema::Header::read, which now becomes the canonical public interface for reading the lock file header.

File: external-crates/move/crates/move-package/src/lock_file/schema.rs [hunk 4]
```diff
@@ -133,32 +133,32 @@ impl ToolchainVersion {
...
-        toml::de::from_str::<Schema<Header>>(contents).context("Deserializing lock header")?;
-
-    if header.version > VERSION {
-        bail!(
-            "Lock file format is too new, expected version {} or below, found {}",
-            VERSION,
-            header.version
-        );
-    }
+    #[allow(clippy::ptr_arg)] // Allowed to avoid interface changes.
```
Comment (amnn) [inline external-crates/move/crates/move-package/src/lock_file/schema.rs:147]: I'm assuming the suggestion is from `&String` to `&str` -- this seems pretty harmless, why not adopt it?

**Review by amnn — APPROVED** [review]
LGTM, thanks for the clean-up @rvantonder !

File: external-crates/move/crates/move-package/src/lock_file/schema.rs (hunk 4 above)
Comment (rvantonder) [reply external-crates/move/crates/move-package/src/lock_file/schema.rs]: done!

**Pull request merged**